benches/benches.rs

```
@@ -0,0 +1,573 @@
+
+    let user_decryption_key_bytes = user_decryption_key
+        .as_bytes()
+        .expect("cannot get bytes from user decryption key");
+    let user_decryption_key_ptr = user_decryption_key_bytes.as_ptr().cast::<i8>();
+    let user_decryption_key_len = user_decryption_key_bytes.len() as i32;
+
+    let mut cache_handle: i32 = 0;
+    unsafe {
+        unwrap_ffi_error(h_aes_create_decryption_cache(
+            &mut cache_handle,
+            user_decryption_key_ptr,
+            user_decryption_key_len,
+        ))
+        .expect("cannot create aes encryption cache");
+    }
+
+    c.bench_function("FFI AES header decryption using cache", |b| {
+        b.iter(|| unsafe {
+            unwrap_ffi_error(h_aes_decrypt_header_using_cache(
+                symmetric_key_ptr,
+                &mut symmetric_key_len,
+                uid_ptr,
+                &mut uid_len,
+                additional_data_ptr,
+                &mut additional_data_len,
+                encrypted_header
+                    .encrypted_header_bytes
+                    .as_ptr()
+                    .cast::<i8>(),
+                encrypted_header.encrypted_header_bytes.len() as c_int,
+                cache_handle,
+            ))
+            .expect("Failed unwrapping FFI AES encrypt header operation")
+        })
+    });
+
+    unsafe {
+        unwrap_ffi_error(h_aes_destroy_decryption_cache(cache_handle))
+            .expect("cannot destroy encryption cache");
+    }
+}
+
+#[cfg(feature = "ffi")]
+unsafe fn unwrap_ffi_error(val: i32) -> anyhow::Result<()> {
+    if val != 0 {
+        let mut message_bytes_key = vec![0u8; 4096];
+        let message_bytes_ptr = message_bytes_key.as_mut_ptr().cast::<i8>();
+        let mut message_bytes_len = message_bytes_key.len() as c_int;
+        get_last_error(message_bytes_ptr, &mut message_bytes_len);
+        let cstr = CStr::from_ptr(message_bytes_ptr);
+        anyhow::bail!("ERROR: {}", cstr.to_str()?);
```

You can return an error with `Err(eyre::eyre(...))`. I am not sure this is better, but this would make the two if branch more similar.

---

`bail!` has been designed to avoid writing `Err(eyre::eyre(...))`, plus this will go away when https://github.com/Cosmian/abe_gpsw/issues/9 is done